### DISCUSSION TITLE
API Auth tokens only last for 8 hours max now? How to refresh programmatically on Teslamate?

### DISCUSSION THREAD

**Discussion by Tesla0** [teslamate-org/teslamate#2606] [Q&A] [answered]
Is it the case that Tesla API Auth tokens now only last for 8 hours? This seems to be the case according to a post from Adrian 10 days ago: [https://github.com/adriankumpf/tesla_auth/issues/34](https://github.com/adriankumpf/tesla_auth/issues/34) . 

Using the link that Adrian posted there, I can refresh the token using a cUrl command, but how can I insert this newly refreshed token into Teslamate programmatically? The only way I currently know how to do this is through the Teslamate UI, by logging out of Teslamate, and then pasting in the Access and Refresh tokens on the login screen. 

Technically, I can probably create a script that does this by mimicking mouse and keyboard movements (or perhaps by using something like Webdriver?), but I was wondering if anyone knows of any better way to paste the refreshed tokens into Teslamate programmatically?

**Comment by cwanja** [comment, marked as answer]
TeslaMate already handles this itself. Tail your TeslaMate logs at startup:
```
Attaching to teslamate_teslamate_1
teslamate_1  | waiting for postgres at database:5432
teslamate_1  | waiting for postgres at database:5432
teslamate_1  | 2022-04-15 11:27:35.328 [info] Migrations already up
teslamate_1  | 2022-04-15 11:27:40.747 [info] System Info: Erlang/OTP 24 (emu)
teslamate_1  | 2022-04-15 11:27:40.747 [info] Version: 1.26.1
teslamate_1  | 2022-04-15 11:27:41.364 [info] POST https://auth.tesla.com/oauth2/v3/token -> 200 (492.820 ms)
teslamate_1  | 2022-04-15 11:27:41.365 [info] Refreshed api tokens
teslamate_1  | 2022-04-15 11:27:41.381 [info] Scheduling token refresh in 6 h
teslamate_1  | 2022-04-15 11:27:41.392 [info] Running TeslaMateWeb.Endpoint with cowboy 2.9.0 at :::4000 (http)
teslamate_1  | 2022-04-15 11:27:41.396 [info] Access TeslaMateWeb.Endpoint at http://localhost
teslamate_1  | 2022-04-15 11:27:41.405 [info] GET /
teslamate_1  | 2022-04-15 11:27:41.748 [info] Starting logger for 'Catalina'
```

Specifically:
**teslamate_1  | 2022-04-15 11:27:41.365 [info] Refreshed api tokens
teslamate_1  | 2022-04-15 11:27:41.381 [info] Scheduling token refresh in 6 h**

**Reply by Tesla0** [reply]
Oh, interesting, I thought the refresh was only for the 90-day "Owner API tokens", which Tesla apparently disabled from the API within the past few days (e.g. "Remove ability to generate Owner API tokens as Tesla does not support them anymore" [source](https://github.com/adriankumpf/tesla_auth/releases/tag/v0.5.0)). Maybe this refresh also applies to the 8-hour token? But then why were the "Owner API tokens" ever needed in the first place? 🤔

**Reply by cwanja** [reply]
Totally speculating, but I think those owner tokens were the legacy way TeslaMate and others were authenticating. Then Tesla changed, TeslaMate upgraded to reflect that change and has been working fine since. 

Edit: do not have the speculate anyways:
https://github.com/adriankumpf/teslamate/releases/tag/v1.25.0
`Use SSO access tokens instead of Owner API tokens (except for Chinese accounts)`

**Reply by Tesla0** [reply]
Ah, awesome, nice find with that release tag! Yeah, I guess that means that the regular 8-hour tokens get auto-refreshed and we never have to manually refresh them anymore. Awesome, good to know, thanks! 👍

**Reply by Tesla0** [reply]
...though some questions still remain, like then how does this explain Adrian's message in [this issue](https://github.com/adriankumpf/tesla_auth/issues/34)?:

[joepalexander](https://github.com/joepalexander) commented [10 days ago](https://github.com/adriankumpf/tesla_auth/issues/34#issuecomment-1088948987)
> @adriankumpf So I guess the best we can now hope for is an 8 hour access token?

[joepalexander](https://github.com/joepalexander) commented [10 days ago](https://github.com/adriankumpf/tesla_auth/issues/34#issuecomment-1088969332)
> Although i guess a way to circumvent this would be to run this headlessly and refresh this on an 7h59m cron job

[adriankumpf](https://github.com/adriankumpf) commented [10 days ago](https://github.com/adriankumpf/tesla_auth/issues/34#issuecomment-1089011472) •
> Sure, it depends on your use case. The Owner API tokens also had to be refreshed at some point, right? Just [refresh your access token](https://tesla-api.timdorr.com/api-basics/authentication#refreshing-an-access-token) every couple of hours.

**Reply by cwanja** [reply]
What specific message? He is just explaining that the Owners API token functionality is being removed from that application. And if you still need it and we're using tesla_auth to programmatically get that token, you will need to find something else.

**Reply by Tesla0** [reply]
Yeah, I think you're right. Looking at the context of the above issue, I don't think it's specific to TeslaMate per se, more of a general way to retrieve auth tokens from the API. Yeah, so I guess it makes sense that TeslaMate probably handles the refresh for you automatically with the 8-hour SSO token. Thanks for your responses, very very helpful!

**Comment by myModelY** [comment]
teslamate is horror now. after 1.27.0 it doesn´t work well. i think i switch to teslalogger
sory

**Reply by denisjoshua** [reply]
I use teslalogger and teslamate... the last one is better :-)

**Reply by myModelY** [reply]
sure? last time teslamate ask every 3 days for token . logging stopped , until new token is input.
many charges, drives are missing, because i got no message if new token is waiting for

**Reply by denisjoshua** [reply]
In the past 2 years, I insert a token maybe 2 times... I think that your settings is wrong at this point.

**Reply by myModelY** [reply]
and where? i only install, setup how it is written here.
first time user/pwd from tesla and ready. which settings did you mean?

**Reply by denisjoshua** [reply]
I don't know... but what I know is that teslamate and teslafi is the unique software that didn't ask me to refresh the key.
